File: home/data/contests/proto_asura/problems/A/extras/wrappers/js/PlayerWrapper.js
load('wrappers/js/IOProcessor.js');

/**
 * Wrapper for players in JavaScript.
 *
 * @author José Carlos Paiva <code>josepaiva94@gmail.com</code>
 */
var PlayerWrapper = function () {
    this.player_id = undefined;
    this.action = {'command': {}, 'messages': []};
};


/********************************************************************************************
 *                         Methods implemented by concrete players                          *
 ********************************************************************************************/

/**
 * Get the name of this player from the concrete player
 *
 * @return {string} the name of this player
 */
PlayerWrapper.prototype.getName = function () {
    return '';
};

/**
 * Provide the concrete player with an opportunity to perform
 * initializations
 */
PlayerWrapper.prototype.init = function () {
};

/**
 * Invoked on the concrete player to ask for his action
 */
PlayerWrapper.prototype.execute = function () {
};


/********************************************************************************************
 *                         Methods implemented by abstract players                          *
 ********************************************************************************************/

/**
 * Update state on abstract players
 *
 * @param state_update {object} the state update
 */
PlayerWrapper.prototype.update = function (state_update) {
};

/**
 * Invoked on the abstract player to manage player lifecycle
 */
PlayerWrapper.prototype.run = function () {
};


/********************************************************************************************
 *                                     General Helpers                                      *
 ********************************************************************************************/

/**
 * Set the player ID
 *
 * @param player_id {string} ID of the player
 */
PlayerWrapper.prototype.setPlayerId = function (player_id) {
    this.player_id = player_id;
};

/**
 * Get the player ID
 *
 * @return {string} ID of the player
 */
PlayerWrapper.prototype.getPlayerId = function () {
    return this.player_id;
};

/**
 * Read a line of input and update the the state of the player with a state update.
 * It expects that the line contains the JSON string representation of a
 * state update.
 */
PlayerWrapper.prototype.readAndUpdate = function () {

    var state_update = JSON.parse(IOProcessor.readln());
    this.update(state_update);
};

/**
 * Send the current action (command and log messages) to the Game Manager
 */
PlayerWrapper.prototype.sendAction = function () {

    var actionToSend = this.action;

    this.action = {'command': {}, 'messages': []};

    IOProcessor.writeln(JSON.stringify(actionToSend));
};

/**
 * Do an action in the game with given name and arguments
 *
 * @param name {string} Name of the command
 * @param args {Array}  Arguments of the command
 */
PlayerWrapper.prototype.doAction = function (name, args) {
    this.action.command = {'name': name, 'args': args};
};

/**
 * Log debugging message
 *
 * @param message debugging message
 */
PlayerWrapper.prototype.log = function (message) {
    this.action.messages.push(message);
};

/**
 * Send the name to the Game Manager
 */
PlayerWrapper.prototype.sendName = function () {
    var name = this.getName();
    this.doAction("NAME", [!name ? this.player_id : name]);
    this.sendAction();
};

/********************************************************************************************
 *                                       Main                                               *
 ********************************************************************************************/

// parsing arguments
if (scriptArgs.length !== 2) {
    printErr('usage: {path/to/player/code} {player-id}');
    quit(1);
}

var program_path = scriptArgs[0];

// initialize player
load(program_path);

var player = new Player();

player.setPlayerId(scriptArgs[1]);

// main
player.init();
player.sendName();
player.run();

File: home/data/contests/proto_asura/problems/C/extras/wrappers/java/PlayerWrapper.java
package wrappers.java;

import com.google.gson.GsonBuilder;

import java.util.Scanner;

/**
 * Wrapper for players in Java
 *
 * @author José Carlos Paiva <code>josepaiva94@gmail.com</code>
 */
public abstract class PlayerWrapper {

    private static Scanner in = new Scanner(System.in);

    private String playerId;
    private PlayerAction action = new PlayerAction();

    private Json json = Json.get();

    /********************************************************************************************
     *                         Methods implemented by concrete players                          *
     ********************************************************************************************/

    /**
     * Get the name of this player from the concrete player
     *
     * @return {@code String} the name of this player
     */
    public abstract String getName();

    /**
     * Provide the concrete player with an opportunity to perform
     * initializations
     */
    public abstract void init();

    /**
     * Invoked on the concrete player to ask for his action
     */
    public abstract void execute();

    /********************************************************************************************
     *                         Methods implemented by abstract players                          *
     ********************************************************************************************/

    /**
     * Update state on abstract players
     *
     * @param {@link StateUpdate} the state update
     */
    public abstract void update(StateUpdate update);

    /**
     * Invoked on the abstract player to manage player lifecycle
     */
    protected abstract void run();

    /********************************************************************************************
     *                                     General Helpers                                      *
     ********************************************************************************************/

    /**
     * Set the player id
     *
     * @param playerId ID of the player
     */
    public void setPlayerId(String playerId) {
        this.playerId = playerId;
    }

    /**
     * Get the player id
     *
     * @return ID of the player
     */
    public String getPlayerId() {
        return playerId;
    }

    /**
     * Read a line of input and update the the state of the player with a {@link StateUpdate}.
     * It expects that the line contains the JSON string representation of a
     * {@link StateUpdate}.
     */
    public void readAndUpdate() {
        String line = in.nextLine();
        StateUpdate stateUpdate = json.objectFromString(line, StateUpdate.class);
        update(stateUpdate);
    }

    /**
     * Send the current action (command and log messages) to the {@code GameManager}
     */
    public void sendAction() {

        PlayerAction action = this.action;

        this.action = new PlayerAction();

        String actionString = json.objectToString(action);
        System.out.println(actionString);
    }

    /**
     * Do an action in the game with given name and arguments
     *
     * @param name Name of the command
     * @param args Arguments of the command
     */
    public void doAction(String name, Object...args) {
        action.setCommand(new Command(name, args));
    }

    /**
     * Log debugging message
     *
     * @param message debugging message
     */
    public void log(String message) {
        action.log(message);
    }

    /**
     * Send the name to the {@code GameManager}
     */
    private void sendName() {
        doAction("NAME", getName());
        sendAction();
    }

    /********************************************************************************************
     *                                      Static helpers                                      *
     ********************************************************************************************/

    public final static void main(String[] args) {

        if (args.length == 2) {
            String playerClassName = args[0];
            String playerId = args[1];

            PlayerWrapper player = null;
            try {
                player = initializePlayer(playerClassName);
            } catch (InstantiationException | IllegalAccessException | ClassNotFoundException e) {
                System.err.println("Error initializing player class " + playerClassName);
                System.exit(9);
            }

            player.setPlayerId(playerId);

            try {
                player.init();

                player.sendName();

                player.run();
            } catch (Exception e) {
                System.err.println("Error while running player: " + e.getMessage());
                System.exit(9);
            }
        } else {
            System.err.println("usage: {player.qualified.ClassName} {player-id}");
        }
    }

    /**
     * Instantiates a {@link PlayerWrapper} from a given binary name
     *
     * @param playerClassName class name that implements {@link PlayerWrapper}
     * @return instance of playerClassName
     * @throws InstantiationException - Failed to get an instance of the player
     * @throws IllegalAccessException - Illegal access to class instance
     * @throws ClassNotFoundException - Player class was not found
     */
    private static PlayerWrapper initializePlayer(String playerClassName)
            throws InstantiationException, IllegalAccessException, ClassNotFoundException {
        ClassLoader loader = ClassLoader.getSystemClassLoader();
        Class<?> clazz = loader.loadClass(playerClassName);
        Object object = clazz.newInstance();

        if (object instanceof PlayerWrapper)
            return (PlayerWrapper) object;
        else
            throw new IllegalArgumentException("Not a player class:" + playerClassName);
    }
}
